Sourcemaps stacktrace tests

puppyofkosh opened the pull request:
I added a few tests for sourcemaps. Most of them are pretty simple, testing calling a non function or non method, or accessing a field that doesn't exist. It also tests errors from within flat functions, to be sure that we get complete stack traces.

When testing the stack trace of a function that's tail recursive, we only check that the top and bottom frames are correct, since the middle "frames" are omitted by the optimization. This might be controversial so any thoughts are welcome.

I also made some changes in anf-loop-compiler. There are a few places where code that might throw an exception was not sourcemapped and so the stack trace would be wrong or incomplete. I added a few j-sourcenodes in places that I thought could use them.

jpolitz commented on src/arr/compiler/anf-loop-compiler.arr:
```diff
@@ -217,12 +217,27 @@ fun obj-of-loc(l):
-  j-app(get-field-loc, [clist: obj, field, loc])
+# Use when we're sure the field will exist
+fun get-field-unsafe(obj :: J.JExpr, field :: J.JExpr, loc-expr :: J.JExpr):
+  j-app(get-field-loc, [clist: obj, field, loc-expr])
```
In this case, I bet we could do `j-dot(j-dot(obj, "dict"), field)`

jpolitz replied:
Oh, I take it back.  Methods and `ref` fields.  My bad.

jpolitz commented on src/js/trove/load-lib.js:
```diff
@@ -97,6 +97,16 @@
+
+      var res = getModuleResultResult(mr);
+      var stackString = runtime.printPyretStack(res.exn.pyretStack);
+      return runtime.makeString(stackString);
```
Would getting back a `raw-array` be more useful, split on new lines?  Just thinking about possible interfaces.

jpolitz commented:
Awesome work.

Two other things I'd like to see here:

- Traces through higher-order functions like `map`
- Traces after bouncing – I realized we're probably missing a case here.  What about calling something that recurses (non-tail) for a while to cause the stack to be collected, _then_ throws an exception.

jpolitz commented:
Also traces through higher-order builtins like `raw-list-map`, which is different from `map` in lists because there is no Pyret code involved at all.

puppyofkosh commented on src/js/base/exn-stack-parser.js:
```diff
@@ -30,9 +30,40 @@ define("pyret-base/js/exn-stack-parser", [], function() {
   }
 
+  function convertExceptionToPyretStackTrace(e, program) {
+    var parsedStack = parseStack(e.stack);
```
This was duplicated in handalone and load-lib

puppyofkosh commented on src/js/base/exn-stack-parser.js:
```diff
@@ -30,9 +30,40 @@ define("pyret-base/js/exn-stack-parser", [], function() {
+      return posForPyret;
+    });
+
+    if (e.pyretStack) {
```
This is the only actual "new" code here. It appends the pyret stack which was converted from a javascript stack (using sourcemaps) to the stack frames which were collected during bounces.

puppyofkosh commented on src/js/base/runtime.js:
```diff
@@ -7,6 +7,7 @@ define("pyret-base/js/runtime",
    "pyret-base/js/secure-loader",
    "seedrandom"],
 function (Namespace, jsnums, codePoint, util, exnStackParser, loader, seedrandom) {
+  Error.stackTraceLimit = Infinity;
```
Had to add this for some of the tests with larger stack traces. If you think there's a better place for it let me know

puppyofkosh commented on tests/pyret/tests/test-repl.arr:
```diff
@@ -138,6 +145,118 @@ check:
+  result46.v satisfies L.is-failure-result
+  stacktrace46 = L.get-result-stacktrace(result46.v)
+
+  raw-array-length(stacktrace46) is 5
```
Here are the new tests which should do what you suggested

puppyofkosh commented:
Okay I added some new tests (the second one you suggested is particularly fun I think. I added some code to "append" the two stacks we have (see comment)).

There was a bit of trouble with the stack trace size limit being small by default, but following your comment in #random I changed it to infinity.
